### ISSUE TITLE
`collections.namedtuple`/`typing.NamedTuple` bug: Customized synthetic constructor being `__init__` instead of `__new__` causes multiple issues when subclassing

### ISSUE THREAD

**Issue by Sachaa-Thanasius** [microsoft/pyright#7769]
**Describe the bug**
`collections.namedtuple` and `typing.NamedTuple` from the standard library provide users with bespoke tuple subclasses with custom constructors and properties to emulate attribute access. Part of this process involves creating a custom `__new__` based on input. However, pyright models the custom constructors by synthesizing corresponding `__init__` methods instead. That incorrect base assumption about namedtuples leads to a whole host of issues when subclassing a namedtuple, since pyright assumes that `__init__` must be overriden instead of `__new__` and doesn't compensate for `__new__` being overriden. As a result, overriding either results in either false positive bugs from pyright or runtime errors from python due to a mismatch in expected vs. actual constructor signatures.

**Code or Screenshots**
For the sake of brevity and showing better synthesized annotations, I've only used `typing.NamedTuple` below, but similar behavior occurs with `collections.namedtuple`. Additionally, I didn't show the result of every reveal_type, though they're visible in this [pyright playground link](https://pyright-play.net/?strict=true&code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwChHRIokBPXLUgc2zwEiAOWIQ0AEwAqCXABs0AGijoAbmmJyYnXGmYBjOcXr0oAIWNoAFKPHTZCgJQAuRlHdRSYtM6gAjMDA5Nw8kAAseXl96JBAQ9zC0OTkwXzksGIBtHiQAXWZGNQ0tHWsLejQAOjhSNAB3OEd3AGIoKS40KDBgKAAicqqa%2Brhe7FNeq0N6X1LMgGUk4AABAdzlACpiEF5pqABBUg4N9YBrOq2d3wOOJoBaAD4oBbllgd7CtHVNbQ6rAeqYDwsEhGi02h0uj1%2BpYAUCQTBRhk%2BlYKi9fANlF5xL4AkFlOFItFYspEslUlB0lkcrk7o9hHQ0O9WrZJDJ5GhTGFiOooPoEDE2EMGjAoOJwmAJPRMZRYIDSMDMGKwhL6JVGK0pIkKlB6FheF4UOhTPowAg5BJ-J16HVgfpEhabeFRTwCKKJVhgFh9NRaAwSKQLcCoHVTebLW70CR9HyQNQqsw0AAPACMUAAvOZLFYpCAEEo%2BgS%2BL1lJlk8oAEzKADMNIKhmMpgAoomxOzy39LC54lAJGgekK4JM5FLPN4cYE5PiInwiSANqSUr4cl2PKuVGhDaQdbI0CArI4AbVhUOR1j84XeMpKUgrAuwI5HAUit9SlZm62FOXD8MYI-nyVfnfPBP1heV4UfRgk3LdMoCAtts1zfNegvYsoDLKBKygKsmigVoAHUCBOUwJFQSIVAQUgaHEZQ-GQKAuVMYgoGATQKlufBdRoHlSluO00H0E4yN3cAQF8ZtdH0JBJCwqAOOBX1NBIbYEHESjVSgOsjBMWCW2AtAqw7CoVw8Xt%2BzlBUYBRRZMTHfwJynQkdWJKB1jvJdKOMtdt10PcDzgOFBzPBy%2BCvDIbzvB8n0%2BYofl0N9dPZKtv2FP9opfQCEoUJL-LAxoE0TKsYLgrKELzZRkOnXhUPQzDsLBAAxYgsGHEgiFzSicHzWiiAodgOl4xIBKEkARLVRggA) and display the mismatch in what the type checker sees and what the runtime is.

```python
from __future__ import annotations

from typing import NamedTuple, reveal_type

class Base(NamedTuple):
    name: bool
    thing: str
    hello: list[int]


reveal_type(Base.__new__)  # Type of "Base.__new__" is "(cls: type[Self@Base], *args: Any, **kwargs: Any) -> Self@Base"
reveal_type(Base.__init__)  # Type of "Base.__init__" is "(self: Base, name: bool, thing: str, hello: list[int]) -> None"
# NamedTuples have custom __new__ methods, not __init__ methods.
# These signatures could be switched with minor modifications and it would be more accurate.

ex1 = Base(True, "thing", [1, 2, 3])


class Example2(Base):
    def __new__(cls, name: bool, thing: str, *hello: int):
        return super().__new__(cls, name, thing, list(hello))


reveal_type(Example2.__new__)
reveal_type(Example2.__init__)

ex2 = Example2(True, "thing", 1, 2, 3)  # Works during runtime, but has a false-positive type-checking error: Expected 3 positional arguments. 


class Example3(Base):
    def __init__(self, name: bool, thing: str, *hello: int):
        super().__init__(name, thing, list(hello))


reveal_type(Example3.__new__)
reveal_type(Example3.__init__)

ex3 = Example3(True, "thing", 1, 2, 3)  # Fails at runtime, but no type-checking error.
#   File "/home/thanos/projects/cfgv/normal.py", line 38, in <module>
#     ex2 = Example3(True, "thing", 1, 2, 3)
#           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
# TypeError: Example1.__new__() takes 4 positional arguments but 6 were given
```

If constructor synthesis within pyright was shifted onto the namedtuple's `__new__` method instead of `__init__`, I believe this behavior would be fixed. However, that's just my guess as a regular pyright user.

**VS Code extension or command-line**
VS Code with Pylance as well as pyright playground (version 1.1.360).

**Comment by erictraut** [comment]
Both the `__new__` and `__init__` methods must be synthesized in the case of a NamedTuple because the base class (`object`) defines these methods such that they accept no parameters. However, I agree that it would be better in the case of NamedTuple to synthesize the `__new__` method based on the field names rather than the `__init__` method. The reason it currently synthesizes the `__init__` method based on field names is that this code is shared with dataclasses.

I've updated the logic in this shared code to conditionally synthesize the `__new__` method for NamedTuple subclasses. This will be included in the next release.

**Comment by Sachaa-Thanasius** [comment]
I think I understand. Thank you!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.361, which I just published.
